compiler/annotation/type_seeder.rs

```diff
@@ -1583,13 +1584,13 @@ impl BinaryConstraint for RelationRoleEdge<'_> {
                 return Ok(());
             } // It can't be another type => Do nothing and let type-inference clean it up
         };
-        relation
-            .get_relates(seeder.snapshot, seeder.type_manager)?
-            .iter()
-            .map(|relates| TypeAnnotation::RoleType(relates.role()))
-            .for_each(|type_| {
-                collector.insert(type_);
-            });
+        for relates in relation.get_relates(seeder.snapshot, seeder.type_manager)?.iter() {
+            let is_write_stage_and_relates_is_abstract = seeder.is_write_stage
+                && relation.is_related_role_type_abstract(seeder.snapshot, seeder.type_manager, relates.role())?;
+            if !is_write_stage_and_relates_is_abstract {
+                collector.insert(TypeAnnotation::RoleType(relates.role()));
+            }
+        }
```

Comment:
Though an abstract super-role would satisfy a match, it is invalid in an insert.

For attributes, this case is handled nicely by type-inference because things are never annotated with abstract types. This doesn't cover roles because they're types.